src/order.c

```diff
@@ -0,0 +1,4084 @@
+ * Doubles
+ *
+ * This uses a combination of 2 ordering algorithms:
+ *
+ * - `dbl_insertion_order()` - An insertion sort is used when `x` is very small.
+ *
+ * - `dbl_radix_order()` - This is similar to `int_radix_order()`, see above,
+ *   but makes a max of 8 passes over the data.
+ *
+ * For doubles, we assume `sizeof(double) == 8`, which should pretty much be
+ * ensured by IEEE 754 specifications.
+ *
+ * For the mapping here, it is possible to map `double -> uint64_t` in an
+ * order preserving way. This is very cool, and involves always flipping the
+ * sign bit of the value, and flipping all other bits if the value was negative.
+ * This is described more in: http://stereopsis.com/radix.html.
+ * This is implemented in `dbl_adjust()` which also handles `na_last` and
+ * `decreasing`. For `na_last`, we treat `NA_real_` and `NaN` equivalently.
+ * Base R does as well, but data.table does not.
+ *
+ * -----------------------------------------------------------------------------
+ * Characters
+ *
+ * Character vector ordering is a bit trickier than integers or doubles. It
+ * uses two internal details in R:
+ *
+ * - CHARSXP values in R are in a global string pool, and
+ *   repeated strings like `c("hi", "hi")` both point to the same CHARSXP.
+ *
+ * - There is a property that all vectors have called the TRUELENGTH. This is
+ *   used to overallocate in lists, but is otherwise unused
+ *   in CHARSXPs. The truelength is stored as a `R_xlen_t`.
+ *
+ * Character ordering is achieved by first iterating through `x` and extracting
+ * the unique values by using the TRUELENGTH. If the truelength is positive
+ * or zero, we save it as an unseen string and set the truelength to `-1`.
+ * Otherwise if it is negative we assume we have seen it already. At the
+ * end of this we have a vector of unique strings. We order this with
+ * `chr_radix_order()`. Then we iterate over the now sorted unique strings
+ * and set their truelength to `-i - 1` (where `i` is the index while
+ * iterating). This marks the unique strings with their ordering (as a negative
+ * value to be different from R) and also updates the original character vector.
+ * We then iterate through the original vector again, plucking off the now
+ * updated truelength integer values. This gives us an integer proxy for the
+ * ordering, which we can run through `int_order_chunk()` to get the final
+ * ordering.
+ *
+ * The ordering of unique characters uses a combination of 2 ordering
+ * algorithms:
+ *
+ * - `chr_insertion_order()` - Used when `x` is small.
+ *
+ * - `chr_radix_order()` - Same principle as integer/double ordering, but
+ *   we iterate 1 character at a time. We assume a C locale here, and any
+ *   non-ASCII and non-UTF8 strings are translated to UTF8.
+ *
+ * -----------------------------------------------------------------------------
+ * Logicals
+ *
+ * Just use the same infrastructure as integers.
```

Comment:
Presumably always be a counting sort?

Reply:
That or an insertion sort for very small vectors. I'll add a note.